qiling/extensions/r2/r2.py

```diff
@@ -0,0 +1,241 @@
+        "w": UC_PROT_WRITE,
+        "x": UC_PROT_EXEC,
+    }
+    for ch in permstr:
+        perm += dic.get(ch, 0)
+    return perm
+
+
+class R2Data:
+    def __init__(self, **kwargs):
+        names = set([f.name for f in fields(self)])
+        for k, v in kwargs.items():
+            if k in names:
+                setattr(self, k, v)
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Section(R2Data):
+    name: str
+    size: int
+    vsize: int
+    paddr: int
+    vaddr: int
+    perm: int
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.perm = perm2uc(self.perm)
+
+
+@dataclass(unsafe_hash=True, init=False)
+class String(R2Data):
+    string: str
+    vaddr: int
+    paddr: int
+    size: int
+    length: int
+    section: str = None
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Symbol(R2Data):
+    # see https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_bin.h
+    SymbolType = Literal["NOTYPE", "OBJ", "FUNC", "FIELD", "IFACE", "METH", "STATIC", "SECT",
+                         "FILE", "COMMON", "TLS", "NUM", "LOOS", "HIOS", "LOPROC", "HIPROC", "SPCL", "UNK"]
+
+    SymbolBind = Literal["LOCAL", "GLOBAL", "WEAK", "NUM", "LOOS", "HIOS", "LOPROC", "HIPROC", "IMPORT", "UNKNOWN"]
+
+    name: str
+    realname: str
+    bind: SymbolBind
+    size: int
+    type: SymbolType
+    vaddr: int
+    paddr: int
+    is_imported: bool
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Function(R2Data):
+    name: str
+    offset: int
+    size: int
+    signature: str
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Flag(R2Data):
+    offset: int
+    name: str = ''
+    size: int = 0
+
+    def __lt__(self, other):
+        return self.offset < other.offset
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Xref(R2Data):
+    XrefType = Literal["NULL", "CODE", "CALL", "DATA", "STRN", "UNKN"]
+
+    name: str
+    fromaddr: int  # from is reserved word in Python
+    type: XrefType
+    perm: int
+    addr: int
+    refname: str
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.fromaddr = kwargs["from"]
+        self.perm = perm2uc(self.perm)
+
+    def __lt__(self, other):
+        return self.fromaddr < other.fromaddr
+
+
+class R2:
+    def __init__(self, ql: Qiling, baseaddr=(1 << 64) - 1, loadaddr=0):
+        super().__init__()
+        self.ql = ql
+        # r2 -B [baddr]   set base address for PIE binaries
+        self.baseaddr = baseaddr
+        self.loadaddr = loadaddr  # r2 -m [addr]    map file at given address
+        self.analyzed = False
+        self._r2c = libr.r_core.r_core_new()
+        if ql.code:
+            self._setup_code(ql.code)
+        else:
+            self._setup_file(ql.path)
+
+    def _qlarch2r(self, archname: str) -> str:
+        archname = archname.lower()
+        # str.removesuffix() is not available until Python 3.9
+        return archname[:-2] if archname.endswith('64') else archname
```

Comment:
Why not do a mapping from enum to r2 arch type?

Reply:
QlArch
```python
class QL_ARCH(IntEnum):
    X86 = 101
    X8664 = 102
    ARM = 103
    ARM64 = 105
    MIPS = 106
    A8086 = 107
    EVM = 108
    CORTEX_M = 109
    RISCV = 110
    RISCV64 = 111
    PPC = 112
```

r2asm
```
[0x00000000]> e asm.arch=
bf            6502          8051          amd29k        arm.as        
arm           arm.winedbg   avr           cr16          dalvik        
dcpu16        gb            h8300         i4004         i8080         
java          lh5801        lm32          m680x         malbolge      
mcore         mcs96         mips          nios2         or1k          
pic           ppc.as        ppc           propeller     riscv         
rsp           sh            snes          sparc         s390          
tms320        tms320c64x    tricore       v810          v850          
vax           ws            x86.as        x86           x86.nasm      
x86.nz        xap           xcore         arm.gnu       cris          
m68k.gnu      lanai         mips.gnu      ppc.gnu       sparc.gnu     
s390.gnu      v850.gnu
```

Only `A8086`, `EVM`, `CORTEX_M` are not seen in r2, the others can all be converted by removing `64` suffix.